File: playwright-python/steps/api_steps.py
from playwright.sync_api import APIResponse
from pytest_bdd import given, parsers, then, when

from clients.storefront_api import StorefrontApi


@when(
    parsers.parse('the client requests "{path}"'),
    target_fixture="api_response",
)
def client_requests_path(
    storefront_api: StorefrontApi,
    path: str,
) -> APIResponse:
    return storefront_api.get(path)


@then("the response is successful")
def response_is_successful(api_response: APIResponse) -> None:
    assert api_response.ok, (
        f"Expected a successful response, got {api_response.status}: "
        f"{api_response.text()}"
    )


@then("it identifies the frontend, products, cart, login, and API exploration paths")
def metadata_identifies_entry_points(api_response: APIResponse) -> None:
    payload = api_response.json()

    assert payload["frontend"] == "/"
    assert payload["products"] == "/products"
    assert payload["cart"] == "/cart"
    assert payload["login"] == "/login"
    assert payload["docs"] == "/products"


@then("each product contains id, name, description, price, stock, image, rating, and review count")
def each_product_contains_expected_fields(api_response: APIResponse) -> None:
    products = api_response.json()
    expected_fields = {
        "id",
        "name",
        "description",
        "price",
        "in_stock",
        "image_url",
        "average_rating",
        "review_count",
    }

    assert isinstance(products, list), "Expected the product response to be a list"
    assert products, "Expected the product collection to contain at least one product"

    for product in products:
        missing_fields = expected_fields - product.keys()
        assert not missing_fields, (
            f"Product {product.get('id', '<unknown>')} is missing fields: "
            f"{sorted(missing_fields)}"
        )


@given("a product with reviews exists")
def product_with_reviews_exists(product_with_reviews: dict[str, object]) -> None:
    assert product_with_reviews["product"].pk is not None
    assert len(product_with_reviews["reviews"]) == 2


@when(
    'the client requests that product from "/products/<id>"',
    target_fixture="api_response",
)
def client_requests_product_detail(
    storefront_api: StorefrontApi,
    product_with_reviews: dict[str, object],
) -> APIResponse:
    product = product_with_reviews["product"]
    return storefront_api.get(f"/products/{product.id}")


@then("the response contains the selected product")
def response_contains_selected_product(
    api_response: APIResponse,
    product_with_reviews: dict[str, object],
) -> None:
    payload = api_response.json()
    product = product_with_reviews["product"]

    assert api_response.ok, api_response.text()
    assert payload["id"] == product.id
    assert payload["name"] == product.name
    assert payload["description"] == product.description
    assert payload["price"] == float(product.price)
    assert payload["in_stock"] == product.in_stock


@then("it contains the product reviews")
def response_contains_product_reviews(
    api_response: APIResponse,
    product_with_reviews: dict[str, object],
) -> None:
    payload = api_response.json()
    reviews_by_user = {review["user"]: review for review in payload["reviews"]}

    assert len(reviews_by_user) == 2
    for expected_review in product_with_reviews["reviews"]:
        actual_review = reviews_by_user[expected_review.user.get_username()]
        assert actual_review["rating"] == str(expected_review.rating)
        assert actual_review["comment"] == expected_review.comment


@when(
    "the client requests an unknown product id",
    target_fixture="api_response",
)
def client_requests_unknown_product(storefront_api: StorefrontApi) -> APIResponse:
    return storefront_api.get("/products/2147483647")


@then(
    parsers.re(r'the API returns status "?(?P<status>\d+)"?'),
    converters={"status": int},
)
def api_returns_status(api_response: APIResponse, status: int) -> None:
    assert api_response.status == status, (
        f"Expected status {status}, got {api_response.status}: {api_response.text()}"
    )


@then('the response says "Product not found"')
def response_says_product_not_found(api_response: APIResponse) -> None:
    assert api_response.json() == {"detail": "Product not found"}


@given("an available product exists")
def available_product_exists(available_product: object) -> None:
    assert available_product.pk is not None
    assert available_product.in_stock > 0


@when(
    'the client posts its id and a valid quantity to "/cart"',
    target_fixture="api_response",
)
def client_adds_product_to_cart(
    storefront_api: StorefrontApi,
    available_product: object,
) -> APIResponse:
    return storefront_api.post(
        "/cart",
        {"product_id": available_product.id, "quantity": 2},
    )


@then("the response contains the updated items, item count, and total")
def response_contains_updated_cart(
    api_response: APIResponse,
    available_product: object,
) -> None:
    payload = api_response.json()

    assert len(payload["items"]) == 1
    assert payload["items"][0]["product"]["id"] == available_product.id
    assert payload["items"][0]["quantity"] == 2
    assert payload["items"][0]["line_total"] == float(available_product.price * 2)
    assert payload["total_items"] == 2
    assert payload["grand_total"] == float(available_product.price * 2)


@when(
    parsers.parse(
        'the client posts product id "{product_id}" and quantity "{quantity}" to "/cart"'
    ),
    target_fixture="api_response",
)
def client_posts_invalid_cart_input(
    storefront_api: StorefrontApi,
    available_product: object,
    product_id: str,
    quantity: str,
) -> APIResponse:
    if product_id == "valid":
        resolved_product_id = available_product.id
    elif product_id == "unknown":
        resolved_product_id = 2147483647
    else:
        resolved_product_id = int(product_id)

    resolved_quantity = (
        available_product.in_stock + 1
        if quantity == "too many"
        else int(quantity)
    )

    return storefront_api.post(
        "/cart",
        {
            "product_id": resolved_product_id,
            "quantity": resolved_quantity,
        },
    )


@given("the cart API contains a product")
def cart_api_contains_product(cart_with_product: object) -> None:
    assert cart_with_product.pk is not None


@when(
    'the client deletes "/cart/<product_id>"',
    target_fixture="api_response",
)
def client_deletes_cart_product(
    storefront_api: StorefrontApi,
    cart_with_product: object,
) -> APIResponse:
    return storefront_api.delete(f"/cart/{cart_with_product.id}")


@then("that product is absent from the returned cart")
def product_is_absent_from_returned_cart(
    api_response: APIResponse,
    cart_with_product: object,
) -> None:
    assert api_response.ok, api_response.text()
    payload = api_response.json()
    returned_product_ids = {
        item["product"]["id"]
        for item in payload["items"]
    }

    assert cart_with_product.id not in returned_product_ids
    assert payload["total_items"] == 0
    assert payload["grand_total"] == 0.0


@given("the cart API contains products")
def cart_api_contains_products(cart_with_product: object) -> None:
    assert cart_with_product.pk is not None


@when(
    'the client deletes "/cart"',
    target_fixture="api_response",
)
def client_clears_cart(storefront_api: StorefrontApi) -> APIResponse:
    return storefront_api.delete("/cart")


@then("the returned cart is empty")
def returned_cart_is_empty(api_response: APIResponse) -> None:
    assert api_response.ok, api_response.text()
    assert api_response.json()["items"] == []


@then("total items and grand total are zero")
def returned_cart_totals_are_zero(api_response: APIResponse) -> None:
    payload = api_response.json()

    assert payload["total_items"] == 0
    assert payload["grand_total"] == 0.0

File: playwright-typescript/src/steps/api-steps.ts
import assert from "node:assert/strict";
import { randomUUID } from "node:crypto";
import { Given, Then, When } from "@cucumber/cucumber";
import type { ApiResponse } from "../support/models.js";
import type { FictoshopWorld } from "../support/world.js";

function response(world: FictoshopWorld): ApiResponse {
  assert.ok(world.state.response, "API response is missing.");
  return world.state.response;
}
function json(world: FictoshopWorld): any { return response(world).json; }
function available(world: FictoshopWorld) { return world.state.products.available; }

When("the client requests {string}", async function (this: FictoshopWorld, path: string) { this.state.response = await this.api.get(path); });
Then("the response is successful", function (this: FictoshopWorld) { assert.ok(response(this).status >= 200 && response(this).status < 300, response(this).text); });
Then("it identifies the frontend, products, cart, login, and API exploration paths", function (this: FictoshopWorld) {
  assert.deepEqual({ frontend: json(this).frontend, products: json(this).products, cart: json(this).cart, login: json(this).login, docs: json(this).docs }, { frontend: "/", products: "/products", cart: "/cart", login: "/login", docs: "/products" });
});
Then("each product contains id, name, description, price, stock, image, rating, and review count", function (this: FictoshopWorld) {
  const products = json(this) as Array<Record<string, unknown>>;
  assert.ok(Array.isArray(products) && products.length > 0);
  for (const product of products) for (const field of ["id", "name", "description", "price", "in_stock", "image_url", "average_rating", "review_count"]) assert.ok(field in product, `Product is missing ${field}`);
});
Given("a product with reviews exists", async function (this: FictoshopWorld) {
  const reference = randomUUID().slice(0, 8);
  const product = await this.data.createProduct(`Reviewed product ${reference}`, "Product created for the API review scenario.", 64.50, 9);
  this.state.products.reviewed = product;
  const one = await this.data.createUser(false);
  const two = await this.data.createUser(false);
  const first = await this.data.createReview(product, one, 4.5, "Excellent test product.");
  const second = await this.data.createReview(product, two, 3.5, "Useful, with room for improvement.");
  this.state.requestedUrls = [`${first.user}|${first.rating}|${first.comment}`, `${second.user}|${second.rating}|${second.comment}`];
});
When("the client requests that product from {string}", async function (this: FictoshopWorld, _path: string) { this.state.response = await this.api.get(`/products/${this.state.products.reviewed.id}`); });
Then("the response contains the selected product", function (this: FictoshopWorld) {
  const actual = json(this);
  const expected = this.state.products.reviewed;
  assert.ok(response(this).status < 300, response(this).text);
  assert.equal(actual.id, expected.id);
  assert.equal(actual.name, expected.name);
  assert.equal(actual.description, expected.description);
  assert.equal(Number(actual.price), expected.price);
  assert.equal(actual.in_stock, expected.in_stock);
});
Then("it contains the product reviews", function (this: FictoshopWorld) {
  const reviews = new Map((json(this).reviews as any[]).map((item) => [item.user, item]));
  assert.equal(reviews.size, 2);
  for (const encoded of this.state.requestedUrls) {
    const [user, rating, comment] = encoded.split("|", 3);
    assert.equal(String(reviews.get(user).rating), Number(rating).toFixed(1));
    assert.equal(reviews.get(user).comment, comment);
  }
});
When("the client requests an unknown product id", async function (this: FictoshopWorld) { this.state.response = await this.api.get("/products/2147483647"); });
Then(/^the API returns status "?(\d+)"?$/, function (this: FictoshopWorld, status: string) { assert.equal(response(this).status, Number(status), response(this).text); });
Then("the response says \"Product not found\"", function (this: FictoshopWorld) { assert.equal(json(this).detail, "Product not found"); });
When("the client posts its id and a valid quantity to {string}", async function (this: FictoshopWorld, path: string) { this.state.response = await this.api.post(path, { product_id: available(this).id, quantity: 2 }); });
Then("the response contains the updated items, item count, and total", function (this: FictoshopWorld) {
  const actual = json(this);
  assert.equal(actual.items.length, 1);
  assert.equal(actual.items[0].product.id, available(this).id);
  assert.equal(actual.items[0].quantity, 2);
  assert.equal(Number(actual.items[0].line_total), available(this).price * 2);
  assert.equal(actual.total_items, 2);
  assert.equal(Number(actual.grand_total), available(this).price * 2);
});
When("the client posts product id {string} and quantity {string} to {string}", async function (this: FictoshopWorld, productId: string, quantity: string, path: string) {
  if (!this.state.products.available) {
    const reference = randomUUID().slice(0, 8);
    this.state.products.available = await this.data.createProduct(`API product ${reference}`, "Product for TypeScript API scenarios.", 49.95, 7);
    await this.api.delete("/cart");
  }
  const resolvedId = productId === "valid" ? available(this).id : productId === "unknown" ? 2147483647 : Number(productId);
  const resolvedQuantity = quantity === "too many" ? available(this).in_stock + 1 : Number(quantity);
  this.state.response = await this.api.post(path, { product_id: resolvedId, quantity: resolvedQuantity });
});
async function populateCartApi(world: FictoshopWorld): Promise<void> {
  const reference = randomUUID().slice(0, 8);
  world.state.products.available = await world.data.createProduct(`API product ${reference}`, "Product for TypeScript API scenarios.", 49.95, 7);
  await world.api.delete("/cart");
  const result = await world.api.post("/cart", { product_id: available(world).id, quantity: 2 });
  assert.equal(result.status, 201, result.text);
}
for (const step of ["the cart API contains a product", "the cart API contains products"]) Given(step, async function (this: FictoshopWorld) { await populateCartApi(this); });
When("the client deletes {string}", async function (this: FictoshopWorld, target: string) {
  const path = target.includes("<product_id>") ? `/cart/${available(this).id}` : target;
  this.state.response = await this.api.delete(path);
});
Then("that product is absent from the returned cart", function (this: FictoshopWorld) {
  assert.ok(response(this).status < 300, response(this).text);
  assert.ok(!(json(this).items as any[]).some((item) => item.product.id === available(this).id));
  assert.equal(json(this).total_items, 0);
  assert.equal(Number(json(this).grand_total), 0);
});
Then("the returned cart is empty", function (this: FictoshopWorld) { assert.ok(response(this).status < 300); assert.equal(json(this).items.length, 0); });
Then("total items and grand total are zero", function (this: FictoshopWorld) { assert.equal(json(this).total_items, 0); assert.equal(Number(json(this).grand_total), 0); });
